File: server/src/main/java/com/sylver/watchwithme/model/RoomState.java
package com.sylver.watchwithme.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;

/**
 * Represents the state of a room of viewers
 */
public class RoomState {

  private final boolean isPlaying;
  private final boolean wasPlaying;
  private final double time;
  private final double prevTime;
  private final int partySize;
  private final String definedBy; // user name

  @JsonCreator
  public RoomState(
    @JsonProperty("isPlaying") final boolean isPlaying,
    @JsonProperty("wasPlaying") final boolean wasPlaying,
    @JsonProperty("time") final double time,
    @JsonProperty("prevTime") final double prevTime,
    @JsonProperty("partySize") final int partySize,
    @JsonProperty("definedBy") final String definedBy
  ) {
    this.isPlaying = isPlaying;
    this.wasPlaying = wasPlaying;
    this.time = time;
    this.prevTime = prevTime;
    this.partySize = partySize;
    this.definedBy = definedBy;
  }

  public boolean getIsPlaying() {
    return isPlaying;
  }

  public boolean getWasPlaying() {
    return wasPlaying;
  }

  public double getTime() {
    return time;
  }

  public double getPrevTime() {
    return prevTime;
  }

  public int getPartySize() {
    return partySize;
  }

  public String getDefinedBy() {
    return definedBy;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    RoomState roomState = (RoomState) o;
    return isPlaying == roomState.isPlaying &&
      wasPlaying == roomState.wasPlaying &&
      Double.compare(roomState.time, time) == 0 &&
      Double.compare(roomState.prevTime, prevTime) == 0 &&
      partySize == roomState.partySize &&
      Objects.equal(definedBy, roomState.definedBy);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(isPlaying, wasPlaying, time, prevTime, partySize, definedBy);
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
      .add("isPlaying", isPlaying)
      .add("wasPlaying", wasPlaying)
      .add("time", time)
      .add("prevTime", prevTime)
      .add("partySize", partySize)
      .add("definedBy", definedBy)
      .toString();
  }
}

File: chrome-extension/src/models/RoomState.js
// SYNC WITH RoomState.java

import _ from 'underscore';

class RoomState {
  constructor(isPlaying, wasPlaying, time, prevTime, partySize, definedBy) {
    this._isPlaying = isPlaying;
    this._wasPlaying = wasPlaying;
    this._time = time;
    this._prevTime = prevTime;
    this._partySize = partySize;
    this._definedBy = definedBy;
  }
  get isPlaying() {
    return this._isPlaying;
  }
  get wasPlaying() {
    return this._wasPlaying;
  }
  get time() {
    return this._time;
  }
  get prevTime() {
    return this._prevTime;
  }
  get partySize() {
    return this._partySize;
  }
  get definedBy() {
    return this._definedBy;
  }

  toJSON() {
    return {
      isPlaying: this._isPlaying,
      wasPlaying: this._wasPlaying,
      time: this._time,
      prevTime: this._prevTime,
      partySize: this._partySize,
      definedBy: this._definedBy,
    };
  }

  static fromJSON(obj) {
    const { isPlaying, wasPlaying, time, prevTime, partySize, definedBy } = obj;
    if (
      _.isBoolean(isPlaying) &&
      _.isBoolean(wasPlaying) &&
      _.isNumber(time) &&
      _.isNumber(prevTime) &&
      _.isNumber(partySize) &&
      _.isString(definedBy)
    ) {
      return new RoomState(isPlaying, wasPlaying, time, prevTime, partySize, definedBy);
    }
  }
}

export default RoomState;
